File: bit-manipulation/sol/single-number-3.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        typedef long long ll;
        ll xr = 0;
        for(auto &it : nums){
            xr = xr ^ it;
        }

        ll x = (xr & (xr-1))^xr; // it will let only the rightmost set bit on.

        ll b1 = 0;  // for 1;
        ll b2 = 0;  // for 0;

        for(auto it : nums){
            if((it&x) == 0){
                // it has rightmost bit off.
                b2 = b2 ^ it;
            }else{
                b1 = b1 ^ it;
            }
        }
        return {(int)b1, (int)b2};
    }
};

File: bit-manipulation/sol/single-number-3.java
// --------------------- OPTIMAL : Bucket Method ------------------
// TC -> N
// SC -> 1

class Solution {
    public int[] singleNumber(int[] nums) {
        int xor = 0;
        for(int a : nums) xor = xor ^ a;

        // extract the right most set bit from the xor.
        int mask = (xor & (xor-1))^xor;

        int bucket1 = 0; // on bit container
        int bucket2 = 0; // off bit container.

        for(int a : nums){
            if((a & mask) == 0){
                bucket1 = bucket1 ^ a;
            }else{
                bucket2 = bucket2 ^ a;
            }
        }

        return new int[]{bucket1 , bucket2};
    }
}
